pilot/api/data.py

```diff
@@ -356,6 +382,83 @@ def resolve_replica(self, fspec, allowed_schemas=None):
 
         return {'surl': surl, 'ddmendpoint': ddmendpoint, 'pfn': replica}
 
+    def resolve_protocol(self, fspec, activity, ddm, allowed_schemas):
+        """
+            Resolve protocols according to allowed schema
+            :param fspec: `FileSpec` instance
+            :param activity: activity name
+            :param ddm: ddm storage data
+            :param allowed_schemas: list of allowed schemas or any if None
+            :return: list of dict(endpoint, path, flavour)
+        """
+
+        if isinstance(activity, basestring):
+            activity = [activity]
+
+        protocols = []
+        for aname in activity:
+            if aname == 'es_events':
+                aname = 'pw'
+            if aname == 'es_events_read':
+                aname = 'pr'
+            protocols = ddm.arprotocols.get(aname)
+            if protocols:
+                break
+
+        allow_protocols = []
+        for schema in allowed_schemas:
+            for pdat in protocols:
+                if schema is None or pdat.get('endpoint', '').startswith("%s://" % schema):
+                    allow_protocols.append(pdat)
+
+        fspec.protocols = allow_protocols
+        return fspec.protocols
+
+    def get_ids_to_ddms(self, ddmconf):
+        """
+           Get a map from storage ids to ddmendpoints.
+           :returns: a dict maps from ids to ddmendpoints.
+        """
+        ids_to_ddms = {}
+        for ddmendpoint in ddmconf:
+            ids_to_ddms[ddmconf[ddmendpoint].pk] = ddmendpoint
+        return ids_to_ddms
+
+    def fix_ddmendpoint_from_storage_id(self, files, activity=None, allowed_schemas=[]):
```

Comment:
For me this is again to much complicated and involves to much extra functions seems to be because of wrong workflow point.

why not to declare data properly from the beginning rather than fix it later, from where you get `storage_id`, why not to set proper ddm here?

Moreover it looks like OS mover specific/depended only

Reply:
Here is an example esmerge job: https://bigpanda.cern.ch/job?pandaid=4001750517
When pilot got a job, it returns with  ('inFiles', 'zip://EventService_premerge_4001371542_14702342-4001371542-14065349203-4019-10.tar,zip://EventService_premerge_4001371542_14702342-4001371542-14065349203-4007-10.tar,....),...,('prodDBlockToken', '9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000,9550/1000'), ...., ('ddmEndPointIn', 'UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK,UKI-SCOTGRID-GLASGOW_DATADISK').

When the job is parsed. The input files are set with ddmendpoint=UKI-SCOTGRID-GLASGOW_DATADISK, with storage_id=9550 and pathconvention=1000. In this case, the ddmendpoint UKI-SCOTGRID-GLASGOW_DATADISK is wrong. That's why I need to use storage_id to reconstruct the ddmendpoint.

OS mover is different. If it's objectstore, it needs to get the keypair from panda to set different env. It also needs to construct url because it's not deterministic. For stagein, the input files are not registered to rucio.

Reply:
so, either `.storage_id` or `.ddmenpoint` is used, but not both, correct?

my point here: to keep workflow untouched and clean for all the cases - pass to the the `transfer` function clean FileSpec values.

I mean rather than patching he base logic by special flags (e.g. `is_user_defined_ddmendpoint`) and select either `.ddmendpoint` or `.storage_id` on fly,  you just need at the level of top workflow do properly create FileSpec entries (the executor or who creates Stage*Client).   the `transfer` function of client should take properly constructed FileSPec entries with overwritten .ddmendpoint value.. At the level of top workflow you know the `job` instance, so that you also know infosys instance and can resolve back .storage_id from the corrected .ddmendpoint to report back to panda. (or cache it from the beginning when you will patch .ddmendpoint from .storage_id)

Reply:
Ok. I can do it.